src/nanoemoji/util.py

```diff
@@ -85,30 +87,120 @@ def file_printer(filename):
+
+
+def _reload(font: ttLib.TTFont, lazy: bool = True):
+    # Stream font to memory and load it back again
     tmp = BytesIO()
-    ttfont.save(tmp)
+    font.save(tmp)
     tmp.seek(0)
-    ttfont2 = ttLib.TTFont(tmp, lazy=False)
-    for tag in ttfont2.keys():
-        table = ttfont2[tag]
+    font2 = ttLib.TTFont(tmp, lazy=lazy)
+
+
+def load_fully(font: Union[Path, ttLib.TTFont]) -> ttLib.TTFont:
+    if isinstance(font, Path):
+        font = ttLib.TTFont(str(font), lazy=False)
+    else:
+        # A TTFont might be opened lazily and some tables only partially decompiled.
+        # If so, reload it
+        if font.lazy:
+            font = _reload(font, lazy=False)
+
+    # one would think lazy=False meant tags would report isLoaded true
+    # one would be wrong, following fails:
+    # _assert_all_keys_loaded(font)
+
+    # OK, for those of you that are STILL lazy just bloody load
+    for tag in font.keys():
+        table = font[tag]
         # cmap is exceptional in that it always loads subtables lazily upon getting
         # their attributes, no matter the value of TTFont.lazy option.
         # TODO: remove this hack once fixed in fonttools upstream
         if tag == "cmap":
             _ = [st.cmap for st in table.tables]
-        ttfont[tag] = table
 
+    _assert_all_keys_loaded(font)
+
+    # post has a cool pattern of raising errors if it has no glyphOrder attribute
+    # font isn't all that useful w/o glyph order so lets confirm we're ok on that front
+    # TODO contemplate whether this is simply a FontTools bug
+    font.getGlyphOrder()
+
+    return font
+
+
+class OtDataPathEntry(NamedTuple):
+    value: Any
+    attr_name: str  # attr of parent to get value
+
+
+OtDataPath = Tuple[OtDataPathEntry, ...]
+
+
+def traverse_ot_data(root, callback: Callable[[OtDataPath], None]) -> Iterable[Any]:
+    visited = set()
+    frontier: List[OtDataPath] = [(OtDataPathEntry(root, ""),)]
+    while frontier:
+        # path is (value, attr_name) tuples. attr_name is attr of parent to get value
+        path = frontier.pop(0)
+        current = path[-1].value
+        visited.add(id(current))
 
-def reorder_glyphs(ttfont: ttLib.TTFont, glyph_order: List[str]):
+        callback(path)
+
+        converters = getattr(current, "converters", [])
+        for converter in converters:
+            is_simple = isinstance(converter, otConverters.SimpleValue)
+            print("  ", getattr(converter, "name", "?nameless?"), converter, is_simple)
+
+            attr_name = getattr(converter, "name", None)
+            if attr_name:
+                raw_value = getattr(current, attr_name, None)
+            else:
+                print("  ", "^has no name")
+                raw_value = None
+
+            if not isinstance(raw_value, list):
+                raw_value = [raw_value]
+
+            for value in raw_value:
+                if (
+                    value is not None
+                    and hasattr(value, "converters")
```

Comment:
you should check that `isinstance(value, BaseTable)`, having an attribute named "converters" may be too broad. This traverse method is meant for trees of BaseTables

Reply:
in https://github.com/fonttools/fonttools/pull/2551/files#diff-fb09acb555593d2fddbfc5bc3bf295f7727e027cdf97d023ab095a02ef2c9436R860, I also added an `iterSubTables` method to iterate over all the BaseTable-children of a given BaseTable (if any) -- I needed that to call ensureDecompiled recursively to fully decompile GPOS/GSUB from a lazy font. You could use that method to extend your traversal frontier.

Reply:
changed the iterSubTables to return a namedtuple(name, value, index) so you better can fit into your traversal

https://github.com/fonttools/fonttools/pull/2551/commits/ab8fc321a76a960322a39f92cf9895066bb52458

I guess I'm gonna merge that PR before it gets too big

Reply:
Now using the new machinery, it's quite helpful.